pkg/kor/namespaces.go

```diff
@@ -0,0 +1,246 @@
+	for _, ignoreType := range ignoreResourceTypes {
+		if resource == ignoreType {
+			return true
+		}
+	}
+	return false
+}
+
+// TODO: refactor using exception list
+func ignorePredefinedResource(gr GenericResource) bool {
+	// Specific list of resources to ignore - resources created in all namespaced by default
+	if gr.GVR.Resource == "configmaps" && gr.GVR.Version == "v1" && gr.NamespacedName.Name == "kube-root-ca.crt" {
+		return true
+	}
+	if gr.GVR.Resource == "serviceaccounts" && gr.GVR.Version == "v1" && gr.NamespacedName.Name == "default" {
+		return true
+	}
+	if gr.GVR.Resource == "events" {
+		return true
+	}
+	return false
+}
+
+func isNamespaceNotEmpty(
+	gvr *schema.GroupVersionResource,
+	unstructuredList *unstructured.UnstructuredList,
+	filterOpts *filters.Options,
+) bool {
+	for _, unstructuredObj := range unstructuredList.Items {
+		gr := GenericResource{
+			GVR: *gvr,
+			NamespacedName: types.NamespacedName{
+				Namespace: unstructuredObj.GetNamespace(),
+				Name:      unstructuredObj.GetName(),
+			},
+		}
+		// Ignore default cluster resources
+		if ignorePredefinedResource(gr) {
+			continue
+		}
+		// User specified resource type ignore list
+		if ignoreResourceType(gr.GVR.Resource, filterOpts.IgnoreResourceTypes) {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+func isErrorOrNamespaceContainsResources(
+	ctx context.Context,
+	clientset kubernetes.Interface,
+	dynamicClient dynamic.Interface,
+	namespace string,
+	filterOpts *filters.Options,
+) (bool, error) {
+	apiResourceLists, err := clientset.Discovery().ServerPreferredNamespacedResources()
+	if err != nil {
+		return true, err
+	}
+
+	// Iterate over all API resources and list instances of each in the specified namespace
+	for _, apiResourceList := range apiResourceLists {
+		for _, apiResource := range apiResourceList.APIResources {
+			gv := strings.Split(apiResourceList.GroupVersion, "/")
+			gvr, err := getGVR(apiResource.Name, gv)
+			if err != nil {
+				return true, err
+			}
+
+			unstructuredList, err := dynamicClient.Resource(*gvr).Namespace(namespace).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				continue
+			}
+
+			if isNamespaceNotEmpty(gvr, unstructuredList, filterOpts) {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func GetUnusedNamespaces(
+	ctx context.Context,
+	filterOpts *filters.Options,
+	clientset kubernetes.Interface,
+	dynamicClient dynamic.Interface,
+	outputFormat string,
+	opts common.Opts,
+) (string, error) {
+	resources := make(map[string]map[string][]ResourceInfo)
+	diff, err := processNamespaces(ctx, clientset, dynamicClient, filterOpts)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to process namespaces: %v\n", err)
+	}
+
+	if len(diff) > 0 {
+		// We consider cluster scope resources in "" (empty string) namespace, as it is common in k8s
+		if resources[""] == nil {
+			resources[""] = make(map[string][]ResourceInfo)
+		}
+		resources[""]["Namespaces"] = diff
+	}
```

Comment:
This part should be rebuilt by implementing the `GroupBy` option which currently doesn't work:
```
$ kor ns --group-by=resource
kor version: vdev

  _  _____  ____
 | |/ / _ \|  _ \
 | ' / | | | |_) |
 | . \ |_| |  _ <
 |_|\_\___/|_| \_\

Unused s:
+---+------------+---------------+
| # | NAMESPACE  | RESOURCE NAME |
+---+------------+---------------+
| 1 | Namespaces | test          |
+---+------------+---------------+

```
Here's an example from `storageclasses.go`:
```
$ kor sc --group-by=resource
kor version: vdev

  _  _____  ____
 | |/ / _ \|  _ \
 | ' / | | | |_) |
 | . \ |_| |  _ <
 |_|\_\___/|_| \_\

Unused StorageClasss:
+---+-----------+---------------+
| # | NAMESPACE | RESOURCE NAME |
+---+-----------+---------------+
| 1 |           | standard      |
+---+-----------+---------------+
```
```suggestion
	switch opts.GroupBy {
	case "namespace":
		resources[""] = make(map[string][]ResourceInfo)
		resources[""]["Namespace"] = diff
	case "resource":
		appendResources(resources, "Namespace", "", diff)
	}
```